### PR TITLE
Use set to avoid adding twice RelatedRecipient with the same broadcas…

### PR DESCRIPTION
**Pull request by norkans7**
…t and contact IDs

### CODE REVIEW

File: temba/msgs/models.py [hunk 1]
```diff
@@ -535,10 +541,15 @@ def send(self, trigger_send=True, message_context=None, response_to=None, status
                     created_on = created_on + timedelta(seconds=1)
 
                 batch = []
-                recipient_batch = []
+                recipient_batch_set = set()
```
Comment (rowanseymour) [inline temba/msgs/models.py]: Problem I see here is that it still blows up if the same contact is in two different batches. I also feel uncomfortable merging any solution until we've replicated the problem in a unit test. Did you check if the unit tests were passing due to deferred constraint checking?

Comment (norkans7) [inline temba/msgs/models.py]: Right, I am trying to have a unit test for that case of different batches

File: temba/msgs/models.py [hunk 2]
```diff
@@ -520,10 +522,17 @@ def send(self, trigger_send=True, message_context=None, response_to=None, status
             if msg:
                 batch.append(msg)
                 # keep track of this URN as a recipient
-                recipient_batch.append(RelatedRecipient(contact_id=msg.contact_id, broadcast_id=self.id))
+                if msg.contact_id not in existing_recipient_contacts:
```
Comment (nicpottier) [inline temba/msgs/models.py]: This is order N on the existing recipient contacts, which can be large for say a Nigeria broadcast. Need a set here.

File: temba/msgs/models.py [hunk 3]
```diff
@@ -453,7 +453,9 @@ def send(self, trigger_send=True, message_context=None, response_to=None, status
 
         # we batch up our SQL calls to speed up the creation of our SMS objects
         batch = []
-        recipient_batch = []
+        recipient_batch_set = set()
+        existing_recipient_contacts = list(
+            BroadcastRecipient.objects.filter(broadcast_id=self.id).values_list('contact_id', flat=True))
```
Comment (rowanseymour) [inline temba/msgs/models.py]: I don't think this is needed - just keep a set outside of the batch loop which records all contacts seen across all batches

File: temba/msgs/models.py [hunk 4]
```diff
@@ -453,7 +453,8 @@ def send(self, trigger_send=True, message_context=None, response_to=None, status
 
         # we batch up our SQL calls to speed up the creation of our SMS objects
         batch = []
-        recipient_batch = []
+        batch_recipients = []
+        existing_recipients = set(BroadcastRecipient.objects.filter(broadcast_id=self.id).values_list('contact_id', flat=True))
```
Comment (rowanseymour) [inline temba/msgs/models.py]: I did comment previously on this that I don't think we don't need to keep refetching within the batch loop - can't we keep a set of all seen contacts outside of the loop?

Comment (nicpottier) [inline temba/msgs/models.py]: This is outside the loop isn't it?

Comment (rowanseymour) [inline temba/msgs/models.py]: Oh right you are

**Pull request merged**